File: THCS/2022_2023/NgheAn_2223/Cau3/ChonBong.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_map>
#include <algorithm>

int main() {
    std::ifstream input("ChonBong.INP");
    std::ofstream output("ChonBong.OUT");

    int n, k;
    input >> n >> k; // Đọc số quả bóng và số màu

    std::vector<int> colors(n);
    std::unordered_map<int, int> color_count; // Đếm số lượng quả bóng theo màu

    // Đọc mã màu và đếm số lượng
    for (int i = 0; i < n; ++i) {
        input >> colors[i];
        color_count[colors[i]]++;
    }

    // Lưu trữ số lượng bóng theo màu vào một vector
    std::vector<int> counts;
    for (const auto& entry : color_count) {
        counts.push_back(entry.second);
    }

    // Nếu chỉ có một màu, không thể lấy được điểm
    if (counts.size() == 1) {
        output << 0 << std::endl;
        return 0;
    }

    // Tính điểm tối đa
    int max_points = 0;
    std::sort(counts.rbegin(), counts.rend()); // Sắp xếp theo số lượng bóng giảm dần

    // Số lần lấy được bóng tối đa
    while (counts.size() > 1) {
        // Lấy số lượng bóng của màu có số lượng tối đa và màu có số lượng tối thiểu
        int min_count = counts.back();
        int max_count = counts.front();

        // Điểm tăng thêm là 2 (mỗi lần lấy ra 2 bóng)
        max_points += 2;

        // Giảm số lượng bóng đã lấy ra
        counts.front()--; // Giảm màu có số lượng tối đa
        counts.back()--;  // Giảm màu có số lượng tối thiểu

        // Loại bỏ màu đã hết bóng
        if (counts.front() == 0) {
            counts.erase(counts.begin());
        }
        if (counts.back() == 0) {
            counts.pop_back();
        }

        // Sắp xếp lại số lượng
        std::sort(counts.rbegin(), counts.rend());
    }

    output << max_points << std::endl;

    input.close();
    output.close();

    return 0;
}

File: THCS/2022_2023/NgheAn_2223/Cau3/ChonBong.py
def max_points(n, k, colors):
    # Đếm số lượng quả bóng theo màu
    color_count = {}
    for color in colors:
        if color in color_count:
            color_count[color] += 1
        else:
            color_count[color] = 1

    # Nếu chỉ có một màu, không thể lấy được điểm
    if len(color_count) == 1:
        return 0

    counts = sorted(color_count.values(), reverse=True)  # Sắp xếp số lượng bóng theo màu giảm dần
    max_points = 0

    # Tính điểm tối đa
    while len(counts) > 1:
        # Lấy số lượng bóng của màu có số lượng tối đa và màu có số lượng tối thiểu
        min_count = counts[-1]
        max_count = counts[0]

        # Điểm tăng thêm là 2 (mỗi lần lấy ra 2 bóng)
        max_points += 2

        # Giảm số lượng bóng đã lấy ra
        counts[0] -= 1  # Giảm màu có số lượng tối đa
        counts[-1] -= 1  # Giảm màu có số lượng tối thiểu

        # Loại bỏ màu đã hết bóng
        if counts[0] == 0:
            counts.pop(0)  # Loại bỏ màu đầu tiên
        if counts[-1] == 0:
            counts.pop()  # Loại bỏ màu cuối cùng

        # Sắp xếp lại số lượng
        counts.sort(reverse=True)

    return max_points


# Đọc dữ liệu từ tệp ChonBong.INP
with open('ChonBong.INP', 'r') as f:
    n, k = map(int, f.readline().split())
    colors = list(map(int, f.readline().split()))

# Tính điểm tối đa
result = max_points(n, k, colors)

# Ghi kết quả vào tệp ChonBong.OUT
with open('ChonBong.OUT', 'w') as f:
    f.write(str(result) + '\n')
